\section{Conclusion}
This paper presents multi-objective evolutionary algorithm biclustering framework using NSGA-II and bicluster internal validation measures as fitness functions.
The MOEA framework can be readily modified to support a number of quality measures. Employing NSGA-II to drive the search for biclusters allows the end user to define their own measure(s) of bicluster quality, or use several existing bicluster quality measures from the literature, as shown.
The empirical evaluation is based on synthetic data generated for gene expression data. A new synthetic data generator is also presented to provide ease of generation of a wide range of trend preserving datasets.
Lastly, a simple enhancement is proposed to improve the robustness of existing recovery and relevance scores.
With the rapid advancement in technology, the type of genomic data being extracted that are available for analysis is changing. For example, array-based methods (that yield gene expression data) are giving way to genome sequencing methods. However, many of the underlying questions that drive the analysis remain the same. The biclustering method presented in this work can be readily modified to address similar questions in the context of new data formats (or types) generated from new technologies. 

%We evaluated three validation measures: Average Spearmanâ€™s Rho
%(ASR), Submatrix Correlation Score (SCS), and Transposed Virtual Error (VE$^T$). 
%\st{The results demonstrate the effectiveness of these functions in discovering meaningful biclusters. The ASR seems to be the best performing function out of all three. We also introduced a methodology for generating more generalized trend preserving bicluster datasets for use in evaluating performance of biclustering methods. Lastly, we presented a more robust definition of recovery and relevance scores to enhance performance evaluation.} {\color{red}{need to rewrite?}}